README.md
# spmjs.io

[![David Status](http://img.shields.io/david/spmjs/spmjs.io.svg?style=flat)](https://david-dm.org/spmjs/spmjs.io)

`spmjs.io` is the distributed packaging server perfectly matching with [spm@3.x](https://github.com/spmjs/spm/tree/master). Now it is rewritten in javascript from [Yuan](https://github.com/spmjs/yuan/)(the precursors), and is faster, more powerful and easier to deploy.

![](https://i.alipayobjects.com/i/localhost/png/201404/2YQxOTYoFp.png)

## Install

```bash
$ git clone git://github.com/spmjs/spmjs.io.git --depth=1
$ cd spmjs.io
$ npm install
```

## Config

```bash
$ cp config/base.default.yaml config/base.yaml
```

Modify `config/base.yaml` as you need.

## Deploy

Start and stop server by a simple command. (For Unix/Linux)

```bash
$ npm start
```

```bash
$ npm stop
```

Then you have a complete package source server which can interact with [spm3.x](https://github.com/spmjs/spm/tree/master) after add the server address to `~/.spm/spmrc-3x`.

```ini
registry = http://127.0.0.1:3000
```

You can set it via `spm config set registry http://your_spm_server.com`.

Also you can use arguments `--registry` or `-r` for each command.

```bash
$ spm install -r http://127.0.0.1:3000
$ spm publish -r http://127.0.0.1:3000
```

Reindex the packages for elastic search.

```bash
$ npm run reindex
```

## elasticsearch

Require Java 7 environment : https://github.com/Homebrew/homebrew/issues/29910

## TODO:

- spm owner [ls|add|rm]


History.md
1.4.1 / 2015-03-19
==================

* 修正 spm 命令行无法登陆的问题。
* 修复 /repository/search 查询不到数据的问题。

1.4.0 / 2015-03-18
==================

* Use nedb to store session，重启应用后用户不需要重新登录。 #91
* 优化 package 页面的样式。
* 修正同步的一些问题。#85
* 增加服务端管理员删除包的功能。#86
* 提高首页加载的性能。#94
* 使用 [kramed](http://npmjs.com/kramed) 作为 markdown 的解析器。#96
* 修正首页下载列表不满的问题。#102
* 配合 spm@3.4.* 对文档进行了大幅更新。
* 在页脚添加 git rev 信息。#122
* 升级 elasticsearch 到 1.4.3 。
* 修正更换 Github 账户名导致的无法登陆 spmjs 的问题。#101 #121
  - 升级前需要运行 `node scripts/fixGithubLoginToId.js` 更新数据，建议事先做好数据备份。

1.3.0 / 2014-12-26
==================

* 对 `publish -f` 和 `unpublish` 进行调整，只允许管理员进行操作。#81
* 增加对于 semver 版本号的下载支持。#81
* 用 [markdown-it](https://github.com/markdown-it/markdown-it) 代替 [marked](https://github.com/chjj/marked/)。
* 用 [pm2](https://github.com/Unitech/PM2/) 代替 forever 。#90
* 将 `elasticsearch.tar.gz` 移出仓库。#88
* [提升了首页的加载性能](https://github.com/spmjs/spmjs.io/commit/573d7d39dc32ae1d7cfcfc2aff872e8220e0f436)。
* 修复解析 md 文件里长 html 字符串时太慢的[问题](https://github.com/spmjs/spm/issues/1067)。
* 简化了 badge 的文案。
* [改进了 `/repository/search` 的输出格式](https://github.com/spmjs/spmjs.io/commit/7d91382dd6696e98c21f4e9713f6272bc5d9506d)。
* 部分依赖升级。

1.2.0 / 2014-09-09
==================

* 源同步方案。#53
* 发布白名单[方案](https://github.com/spmjs/spmjs.io/commit/3a16e98a70b3b642344eb99a0df8b6eb896b1afe)。
* 增加搜索框提示功能。
* 修复了一个 unpublish 导致模块依赖信息出错的问题。https://github.com/spmjs/spm/issues/948
* 一些样式、图标、信息架构上的优化。

1.1.2 / 2014-07-17
==================

* elasticsearch 升级到 1.3.0，修复命令执行漏洞。http://news.mydrivers.com/1/306/306326.htm
* github 账户权限修改为只读。#52

1.1.1 / 2014-07-17
==================

* 默认配置从 base.yaml 移到 base.default.yaml，部署前需要拷贝
* 修复 404 页面报错的问题

1.1.0 / 2014-07-16
==================

  * 增加今日下载次数和最近下载列表。
  * 增加了一个获取所有模块详细信息的接口。#44
  * 文档托管服务使用 [serve-spm](https://github.com/spmjs/serve-spm) 进行重构。#43
  * 更新了 badge 为扁平化样式。
  * 更新了默认的 base.yaml 配置。
  * 更新了底部UI。
  * 更新了部分文档。
  * 修复搜索结果排序的问题。#42
  * 修复最新版本号排序错误的问题。
  * 修复一个 repository 格式导致的报错。#29
  * 修复大量文案错误。

1.0.0 / 2014-06-10
==================

  * Initial release



app.js

/**
 * Module dependencies.
 */

var express = require('express');
var http = require('http');
var path = require('path');
var multipart = require('connect-multiparty');
var multipartMiddleware = multipart();
var favicon = require('serve-favicon');
var morgan = require('morgan');
var bodyParser = require('body-parser');
var cookieParser = require('cookie-parser');
var session = require('express-session');
var NedbStore = require('connect-nedb-session-two')(session);
var serveStatic = require('serve-static');
var fs = require('fs-extra');
var spmjsioVersion = require('./package').version;
var gitRev = require('git-rev-sync').short();

// load global config
var yaml = require('node-yaml-config');
var CONFIG = yaml.load('./config/base.yaml');
global.CONFIG = CONFIG;

// start sync
require('./sync');

// mkdir data directory
if (!fs.existsSync(CONFIG.wwwroot)) {
  fs.mkdirSync(CONFIG.wwwroot);
}
if (!fs.existsSync(path.join(CONFIG.wwwroot, 'db'))) {
  fs.mkdirSync(path.join(CONFIG.wwwroot, 'db'));
}
if (!fs.existsSync(path.join(CONFIG.wwwroot, 'docs'))) {
  fs.mkdirSync(path.join(CONFIG.wwwroot, 'docs'));
}
if (!fs.existsSync(path.join(CONFIG.wwwroot, 'repository'))) {
  fs.mkdirSync(path.join(CONFIG.wwwroot, 'repository'));
}

// start index cache
require('./lib/cacheIndex')();

var routes = require('./routes');
var account = require('./routes/account');
var repository = require('./routes/repository');
var docs = require('./routes/docs');

var app = express();

// all environments
app.set('port', CONFIG.server.port || 3000);
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'ejs');
app.use(favicon(__dirname + '/public/favicon.ico'));
app.use(morgan('combined'));
app.use(require('./middlewares/raw-body')({
  contentTypes: ['application/x-tar'],
  limit: '50mb'
}));
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());
app.use(cookieParser('spmjs'));
app.use(session({
  secret: 'keyboard cat',
  resave: false,
  saveUninitialized: true,
  store: new NedbStore({
    filename: path.join(__dirname, './data/db/session.db'),
    clearInterval: 24 * 3600 * 1000
  })
}));
app.use(serveStatic(path.join(__dirname, 'public')));

// development only
if ('development' == app.get('env')) {
  app.use(require('errorhandler')());
}

app.get('/', routes.index);
app.get('/search', routes.search);
app.get('/suggest', routes.suggest);
app.get('/packages', routes.all);
app.get('/package/:name', routes.project);
app.get('/package/:name/:version', routes.package);
app.get('/badge/:name', routes.badge);
app.get('/documentation', routes.documentation);
app.get('/documentation/:title', routes.documentation);

app.get('/login', account.login);
app.get('/callback', account.callback);
app.get('/logout', account.logout);
app.get('/account', account.index);
app.post('/account/login', account.authorize);
app.get('/account/:user', account.user);
app.post('/ownership', account.ownership);
app.delete('/ownership', account.ownership);

app.get('/repository', repository.index);
app.post('/repository/upload', multipartMiddleware, repository.package.checkPermission, repository.upload);
app.get('/repository/search', repository.search);
app.get('/repository/since', repository.since);
app.get('/repository/:name', repository.project.get);
app.delete('/repository/:name', repository.package.checkPermission, repository.project.delete);
app.get('/repository/:name/:version', repository.package.get);
app.post('/repository/:name/:version', repository.package.checkPermission, repository.package.post);
app.put('/repository/:name/:version', repository.package.put);
app.delete('/repository/:name/:version', repository.package.checkPermission, repository.package.delete);
app.get('/repository/:name/:version/:filename', repository.filename.get);

app.get('/repositories', repository.data);

app.get('/docs/:name/:version/*', docs);

// 404
app.get('*', function(req, res) {
  res.status(404).render('404.ejs', {
    title: 'No Found - ' + CONFIG.website.title,
    spmjsioVersion: spmjsioVersion,
    gitRev: gitRev,
    anonymous: CONFIG.authorize.type === 'anonymous',
    user: req.session.user,
    GA: CONFIG.website.GA
  });
});

http.createServer(app).listen(app.get('port'), function() {
  console.log('Express server listening on port ' + app.get('port'));
});
